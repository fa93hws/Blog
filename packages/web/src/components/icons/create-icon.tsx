import * as React from 'react';
import * as classnames from 'classnames';
import styles from './icons.css';

export type Size = 'small' | 'medium' | 'large';
type IconProps = {
  size: Size;
  svgContent: string;
  className?: string;
};

const sizeRatioMap: Record<Size, number> = {
  small: 1,
  medium: 2,
  large: 3,
};

export function getWidthFromSize(size: Size) {
  return 8 * 2 * sizeRatioMap[size];
}

function createSvgString(svgContent: string, size: Size) {
  const ratio = sizeRatioMap[size];
  const width = getWidthFromSize(size);
  const openTag = `<svg version="1.1" xmlns="http://www.w3.org/2000/svg" width="${width}" height="${width}" viewBox="0 0 ${width} ${width}"><g transform="scale(${ratio})">`;
  return `${openTag}${svgContent}</g></svg>`;
}

const BaseIcon = ({ size, svgContent, className }: IconProps) => {
  const svg = createSvgString(svgContent, size);
  return (
    <i
      aria-hidden="true"
      className={classnames(styles.iconSpan, className)}
      dangerouslySetInnerHTML={{ __html: svg }}
    />
  );
};

export function createIcon(svgContent: string) {
  const Icon = ({ size, className }: Omit<IconProps, 'svgContent'>) => (
    <BaseIcon size={size} svgContent={svgContent} className={className} />
  );
  return Icon;
}
