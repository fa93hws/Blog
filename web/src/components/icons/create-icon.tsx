import * as React from 'react';
import styles from './icons.css';

export type Size = 'small' | 'medium' | 'large';
type IconProps = {
  size: Size;
  svgContent: string;
  className?: string;
};

const sizeRatioMap: Record<Size, number> = {
  small: 1,
  medium: 2,
  large: 3,
};

export function getWidthFromSize(size: Size) {
  return 8 * 2 * sizeRatioMap[size];
}

function createSvgString(svgContent: string, size: Size) {
  const ratio = sizeRatioMap[size];
  const width = getWidthFromSize(size);
  const openTag = `<svg version="1.1" xmlns="http://www.w3.org/2000/svg" width="${width}" height="${width}" viewBox="0 0 ${width} ${width}"><g transform="scale(${ratio})">`;
  return `${openTag}${svgContent}</g></svg>`;
}

const BaseIcon = ({ size, svgContent }: IconProps) => {
  const svg = createSvgString(svgContent, size);
  return (
    <span
      aria-hidden="true"
      className={styles.iconSpan}
      dangerouslySetInnerHTML={{ __html: svg }}
    />
  );
};

export function createIcon(svgContent: string) {
  const Icon = ({ size }: Omit<IconProps, 'svgContent'>) => (
    <BaseIcon size={size} svgContent={svgContent} />
  );
  return Icon;
}
