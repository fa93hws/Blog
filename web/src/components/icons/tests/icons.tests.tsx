import * as React from 'react';
import { createIcon } from '../create-icon';

describe('createIcon', () => {
  const svg = '<path></path>';
  const Icon = createIcon(svg);

  it('should generate the small icon correctly', () => {
    expect(
      <Icon size="small" className="small-class" title="small-title" />,
    ).toMatchRenderedSnapshot();
  });

  it('should generate the medium icon correctly', () => {
    expect(
      <Icon size="medium" className="medium-class" title="medium-title" />,
    ).toMatchRenderedSnapshot();
  });

  it('should generate the large icon correctly', () => {
    expect(
      <Icon size="large" className="large-class" title="large-title" />,
    ).toMatchRenderedSnapshot();
  });
});
