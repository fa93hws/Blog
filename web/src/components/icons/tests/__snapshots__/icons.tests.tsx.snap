// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createIcon should generate the large icon correctly 1`] = `
<span
  aria-hidden="true"
  class="large-class"
  title="large-title"
>
  <svg
    height="48"
    version="1.1"
    viewBox="0 0 48 48"
    width="48"
    xmlns="http://www.w3.org/2000/svg"
  >
    <g
      transform="scale(3)"
    >
      <path />
    </g>
  </svg>
</span>
`;

exports[`createIcon should generate the medium icon correctly 1`] = `
<span
  aria-hidden="true"
  class="medium-class"
  title="medium-title"
>
  <svg
    height="32"
    version="1.1"
    viewBox="0 0 32 32"
    width="32"
    xmlns="http://www.w3.org/2000/svg"
  >
    <g
      transform="scale(2)"
    >
      <path />
    </g>
  </svg>
</span>
`;

exports[`createIcon should generate the small icon correctly 1`] = `
<span
  aria-hidden="true"
  class="small-class"
  title="small-title"
>
  <svg
    height="16"
    version="1.1"
    viewBox="0 0 16 16"
    width="16"
    xmlns="http://www.w3.org/2000/svg"
  >
    <g
      transform="scale(1)"
    >
      <path />
    </g>
  </svg>
</span>
`;
